### ISSUE TITLE
Make voting optional

### ISSUE THREAD

**Issue by adam111316** [Razzeee/script.trakt#107]
I personally don't rate shows on trakt, so it would be great if the voting dialog box that pops up after an episode is optional.

**Comment by ghost** [comment]
Hi, thanks for this PR, I'm going to test it thoroughly.
Found an error already when re-rating item, rating dialog is displayed but rating is not possible - focus of selected rating and text description is missing and I can only exit the dialog.
<code>
20:47:40 T:2854550336   ERROR: EXCEPTION: Non-Existent Control 11041
20:47:40 T:2854550336   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: &lt;type 'exceptions.RuntimeError'&gt;
                                            Error Contents: Non-Existent Control 11041
                                            Traceback (most recent call last):
                                              File "/home/xbmc/.xbmc/addons/script.trakt/rating.py", line 184, in onInit
                                                self.setFocus(self.getControl(rateID))
                                            RuntimeError: Non-Existent Control 11041
                                            -->End of Python script error report<--
</code>

**Comment by nate1280** [comment]
Woops, I know whats going on there, fix coming shortly

**Comment by nate1280** [comment]
Above error should be fixed.

**Comment by ghost** [comment]
working now, thanks. 
It's currently not possible to unrate items, right?

**Comment by nate1280** [comment]
Not at the moment, no.

Its something that could probably be added.

**Comment by nate1280** [comment]
Looking through the APIs I don't see a way to unrate an item.

**Comment by rectifyer** [comment]
It's the same API call on trakt, you just send "unrate" or "0" to unrate it.

**Comment by nate1280** [comment]
Ok, then I can add action=unrate while this is still open, or as another PR after its closed.  Pretty sure I have it worked out how to handle it.

**Comment by ghost** [comment]
Separate action is a good idea. 
I wouldn't also mind to be able unrate the movie same way as on trakt webpage by selecting previous rating.
Another idea I have is being able to send specific rating to script by another parameter without the need of displaying the rating dialog at all. It could be useful for external scripts.

<code>RunScript(script.trakt,action=rate,media_type=&lt;show|episode|movie&gt;,dbid=#,rating=#)</code>

On the other note I got this error when submitting rating for episode after scrobble: Same dialog error, different id.

<code>13:35:58 T:2855332672   ERROR: EXCEPTION: Non-Existent Control 11029
13:35:58 T:2855332672   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: &lt;type 'exceptions.RuntimeError'&gt;
                                            Error Contents: Non-Existent Control 11029
                                            Traceback (most recent call last):
                                              File "/home/xbmc/.xbmc/addons/script.trakt/rating.py", line 184, in onInit
                                                self.setFocus(self.getControl(rateID))
                                            RuntimeError: Non-Existent Control 11029
                                            -->End of Python script error report<--
</code>

**Comment by nate1280** [comment]
The above error should be fixed by the latest commit.

Todo:
- [x] Add unrate action
- [x] Selecting same rating unrates media, same behaviour as website
- [x] Add rating parameter to rate directly from external source

Those 3 additions I'll work on through out the day.

**Comment by nate1280** [comment]
Unrating features are added now.

So, selecting the same rating during a re-rate will unrate an item, same as it does on the website.

And there is a new action, unrate, which will unrate an item if its rated.  Same optional parameters as the rate action.

As for adding a rating parameter, the only potential issue I see with this is, other plugins/skins/etc that might use this won't know if simple or advanced rating is being used.

**Comment by nate1280** [comment]
@rectifyer once this is done, if its going to be merged, I would like to squash it to one commit before you do.

**Comment by ghost** [comment]
thanks again :)
Note to unrating/re-rating a movie through rating dialog: Whenever current rating is focused, it should have text description changed to reflect the action dialog is going to use if you submit this rating (like it's on trakt site - 'Unrate this movie' instead of 'Great')

**Comment by rectifyer** [comment]
@nate1280 yes, planning to merge this in for 2.3

**Comment by nate1280** [comment]
@ezechiel1917 focused string in the rating dialog now shows unrate this... like it does on the website if its going to unrate the item.

**Comment by nate1280** [comment]
So, I'm looking at whats needed to add the rating parameter, but, I can't really figure out what would use this.

Aside from the fact that there are 2 rating types, how would other plugins know which to use.  Also, the rating only works on library items, so, unless someone wants to have a set key to always rate an item at a certain value I can't see it being used (for now at least).

So, for the moment, going to shelve the rating parameter, and I will be going over the code in this pull request, seeing if there's anything that needs altering/tweaking, and squash the commit.  After this, I think it will be ready barring any major bugs.

**Comment by rectifyer** [comment]
The rating APIs can always accept a 1-10 scale. We default to the 10 heart scale so most people probably have it by now.

**Comment by nate1280** [comment]
If rating is passed, should it be able to cause an unrate?

So, if rating is restricted to the numerical scale, I guess it could work.

**Comment by rectifyer** [comment]
It will only unrate if 0 or "unrate" is sent.

**Comment by N3MIS15** [comment]
nice @nate1280 I haven't had a chance to test yet but anything that gets trakt more intergrated into XBMC is a win in my book ;)

**Comment by nate1280** [comment]
@rectifyer yes, you mentioned about the 0 and unrate earlier,  I meant if a rating is passed should it be allowed to cause an unrate if it's the same rating, or just do nothing. The values allowed for the rating parameter will be restricted from 1 to 10 inclusive

**Comment by N3MIS15** [comment]
If you are asking what to do when the user backs out of the dialog, I would say just cancel, don't change anything.
If you are asking about clicking 10 if the rating is already 10, then yes unrate (this is the way the site works).

**Comment by nate1280** [comment]
It already does what you mentioned. 

I'm referring to if the rating is passed in a parameter (code hasn't been pushed yet, it's sitting waiting atm) how it should behave.

Currently it'll just do nothing if it's the same, which I think is the best option. Since there is no visual feedback that you might be causing an unrate.

I'll be pushing the code for the parameterized rating tomorrow morning, then doing a few passes over the code to check for anything that could be streamlined,  then a squash and I think it'll be ready. Already have my next enhancement thought out, just relies on parts of this addition.

**Comment by N3MIS15** [comment]
if the user cannot see "unrate" then imo an unrate should not happen so yea, do nothing :)

**Comment by ghost** [comment]
I agree with David, sending specific rating as a parameter should only unrate when rating parameter is 0.

**Comment by nate1280** [comment]
Direct rating via a parameter is added now, and it only supports the advanced rating values.

Just add rating=# to an action=rate. Where # is between 1 to 10 inclusive.  To unrate use action=unrate, not rating=0.

I've also changed how data is passed to the service, using a JSON string instead of delimited, its cleaner and easier to work with now.

**Comment by nate1280** [comment]
I've squashed this commit, and gone over it a bit and tweaked a few things here and there.  But as it sits, I believe its ready.

**Comment by N3MIS15** [comment]
any chance of adding a "remoteid" (imdb, tmdb, tvdb) arg so plugins can take advantage of this feature?

**Comment by nate1280** [comment]
All the coding as it sits, relies on library information to do its thing.

Its probably possible, but as it sits now, I'd say leave this addition as it is, and consider non-library items as a possible future enhancement, if its really wanted before merge, I can look into getting it in.  Would need season/episode as well if its an episode thats being rated.

**Comment by nate1280** [comment]
Ignore my previous comment. 

Took a quick look at adding manual rating of non-library items and it was simpler then I thought it would be. I have code in place to support it,  will look at it again tomorrow morning and commit it shortly after. 

Hopefully after this commit it will finally be ready.

**Comment by nate1280** [comment]
Added ability to rate/un-rate non-library item.

Usage to rate an item:

```
RunScript(script.trakt,action=rate,media_type=<movie|show>,remoteid=#)
RunScript(script.trakt,action=rate,media_type=episode,remoteid=#,season=#,episode=#)
```

The above usage can also take the rating=# parameter to set a specific rating. This value can be from 1 to 10 inclusive.

Usage to unrate an item:

```
RunScript(script.trakt,action=unrate,media_type=<movie|show>,remoteid=#)
RunScript(script.trakt,action=unrate,media_type=episode,remoteid=#,season=#,episode=#)
```

Also fixed a small typo when checking if it should rate a specific item, was looking for id, not dbid as per usage.
And I added a Debug line for when the rating dialog closes, just so its known.

**Comment by N3MIS15** [comment]
Thanks, remote rating works great @nate1280. Here is a little plugin i made to test it out if your interested.
https://gist.github.com/N3MIS15/5478681

can rate show, episode and movie via context menu. (xbmcswift is required)

**Comment by nate1280** [comment]
Has anyone thought the notifications that get displayed don't last long enough?

Also, have others been testing this out? Would really like feedback.

**Comment by ghost** [comment]
I have noticed it. I think 5s is fine, however you can prioritize rating status before title in notification. Since site error strings are completely different when there is problem with connection or something, it should be obvious that rating was submitted sucessfully even without displaying text string for it.

Here are notification examples for different rating actions I'd like to suggest:

<code>trakt.tv Rating
Rated 7: The Matrix (1999)</code>

<code>trakt.tv Rating
Already Rated: The Matrix (1999)</code>

<code>trakt.tv Rating
Unrated: The Matrix (1999)</code>

**Comment by N3MIS15** [comment]
"trakt.tv Rating" is pointless as a header really, you have the icon afterall. I think the header should be "Success" or "Error". +1 for @ezechiel1917 idea for message text.

**Comment by nate1280** [comment]
Since there's 2 parts to the notification, I could always make use of this.

Have the title show where it says trakt.tv Rating, and the status (Rated, Unrated, etc) where its shown usually.

**Comment by nate1280** [comment]
Just pushed a different notification style, the one I mentioned.

I use the Ace skin, so how it shows up for me is on the upper left side, it says the title, and on the upper right side the status.

Give it a try and let me know if it seems better.

**Comment by ghost** [comment]
Yeah, that's the other way I'd do it. I just preferred keeping trakt.tv text there, but I guess trakt icon could be enough to identify what's going on.

**Comment by nate1280** [comment]
Well, from the right hand side, it would be obvious that its a trakt notification, unless you have other plugins that do ratings.

It just seems cleaner, instead of using semicolons to prioritize the status.

**Comment by ghost** [comment]
I know, it's just common to have some kind of addon identification (name) in first row of notification (left side in your skin), but ofc you are not limited by that when creating notification.... as long as you have unique icon in notification it should be fine for user :)

**Comment by nate1280** [comment]
Could always swap the two, so the status shows on the left, and add trakt in the text somewhere.

Hoping to get this notification stuff finalized soon, would like to finish this addition off.

**Comment by rectifyer** [comment]
The icon is enough to identify it as trakt imo. Just let me know when it's solid and I will merge it in.

**Comment by nate1280** [comment]
Not all skins use the icon in notifications unfortunately.

But, if its all good as it sits (notification wise), i'll sqaush it for merging.

**Comment by nate1280** [comment]
Its been squashed, don't believe there are any glaring issues, so I think its ready for merge.

Notification can always be adjusted later as well.

**Comment by ghost** [comment]
@nate1280 
I just noticed when there is "," in title name it wraps the text after it to second part of notification and then status is not displayed.

Also I think it might really turn out better to swap the order of both parts like you mentioned, because first part text doesn't scroll. It doesn't look good when there is title text cut which doesn't scroll. Another improvement could be shorten the status text by removing media type and then you might add trakt.tv back there :)

Currently:
<code>The Simpsons - S01E01 - Simpsons Roastin...
Rating for Episode was successfully submitted</code> 

Swapped (second part can scroll):
<code>trakt.tv - Rating submitted successfully
The Simpsons S01E01 - Simpsons Roasting on an Open Fire</code>

<code>trakt.tv - Already rated
The Simpsons S01E01 - Simpsons Roasting on an Open Fire</code>

<code>trakt.tv - Unrated successfully
The Simpsons S01E01 - Simpsons Roasting on an Open Fire</code>

<code>trakt.tv - Already has the same rating
The Simpsons S01E01 - Simpsons Roasting on an Open Fire</code>

What do you think?

**Comment by N3MIS15** [comment]
@ezechiel1917 no need for trakt.tv, just trakt would be fine. Trakt is the name of the site (and the addon)
Maybe @rectifyer was too cheap to get a dot com.. 
jk :P

**Comment by nate1280** [comment]
That would probably work, if others think this as well, will make the change and submit it.

**Comment by bigmong** [comment]
You can just turn off rating.
Trakt > Settings > Ratings > Un-select the top 3 boxes :)

**Comment by Razzeee** [comment]
@adam111316 just deactivate it in the settings ;)

**Comment by adam111316** [comment]
Oh man, I didn't realize he 'rating' tab was there as it required me to scroll across. I've been using trakt for how long and I never noticed haha... Thanks guys :)
